config/initializers/rack_attack.rb

```diff
@@ -17,15 +17,37 @@ class Rack::Attack
   #
   # Key: "rack::attack:#{Time.now.to_i/:period}:logins/ip:#{req.ip}"
   protected_paths = [
-    "/users",              # sign up
-    "/session",            # sign in
-    "/passwords",          # forgot password
-    "/email_confirmations" # resend email confirmation
+    "/users",               # sign up
+    "/session",             # sign in
+    "/passwords",           # forgot password
+    "/email_confirmations", # resend email confirmation
+    "/session/mfa_create"   # mfa sign in
   ]
-  paths_regex = Regexp.union(protected_paths.map { |path| /\A#{Regexp.escape(path)}\z/ })
 
-  throttle('clearance/ip', limit: REQUEST_LIMIT, period: LIMIT_PERIOD) do |req|
-    req.ip if req.path =~ paths_regex && req.post?
+  mfa_forgot_password_regex = Regexp.new("\/users\/\\d+\/password\/mfa_edit")
+  paths_regex = Regexp.union(protected_paths.map { |path| /\A#{Regexp.escape(path)}\z/ }.append(mfa_forgot_password_regex))
+
+  # 100 req in 10 min
+  # 200 req in 100 min
+  # 300 req in 1000 min (0.7 days)
+  # 400 req in 10000 min (6.9 days)
+  (1..4).each do |level|
+    throttle("clearance/ip/#{level}", limit: REQUEST_LIMIT * level, period: (LIMIT_PERIOD**level).seconds) do |req|
+      req.ip if req.path =~ paths_regex && req.post?
+    end
+  end
+
+  protected_api_paths = [
+    "/api/v1/gems/yank",  # gem yank
+    "/api/v1/gems"        # gem push
+  ]
+  add_owner_regex = Regexp.new("\/api\/v1\/gems\/\\w+\/owners")
```

Comment:
The `\w` (word character `[a-zA-Z0-9_]`) in the `add_owner_regex` might be a bit too strict. Here's the route:

```
POST
/api/v1/gems/:rubygem_id/owners(.:format)
api/v1/owners#create {:format=>/json|yaml/, :rubygem_id=>/[A-Za-z0-9\.\-_]+?/}
```

The pattern for `rubygem_id` is a bit more relaxed than `\w+`. Paths with `rubygem_id` values containing `.`, `-` would bypass the throttle.

Consider using `Rails.application.routes.recognize_path(req.path)` to reuse its pattern matching here to prevent this category of issues. 

`recognize_path` returns the controller and action for matching and is less brittle than duplicating and matching on route patterns.

Reply:
Thanks @eliotsykes :100: 
I have updated it to use `recognize_path`.